In [1]:
import numpy as np

In [2]:
myarr = np.array([2,5,9,8])

In [3]:
myarr

array([2, 5, 9, 8])

In [6]:
myarr = np.array([2,5,6,7,11],np.int8)

In [7]:
myarr

array([ 2,  5,  6,  7, 11], dtype=int8)

In [8]:
myarr = np.array([2,3,4444445,32],np.int8)

In [9]:
myarr

array([ 2,  3, 29, 32], dtype=int8)

since our dtype was 8 bit integer so we were not able to store this big number and hence this error occured 

other dtypes that we can use are int32 and int64

In [10]:
myarr[2]

29

In [11]:
myarr[0]

2

In [12]:
myarr = np.array([[2,4,8,11]])

In [13]:
myarr[0]

array([ 2,  4,  8, 11])

In [14]:
myarr[0,1]

4

In [16]:
myarr.shape

(1, 4)

In [18]:
myarr.dtype

dtype('int32')

In [19]:
myarr[0,1] = 45

In [20]:
myarr

array([[ 2, 45,  8, 11]])

In [21]:
listarray = np.array([[2,4,5],[8,9,11],[22,4,1]])

In [22]:
listarray

array([[ 2,  4,  5],
       [ 8,  9, 11],
       [22,  4,  1]])

In [23]:
listarray.dtype

dtype('int32')

In [24]:
listarray.size

9

In [26]:
listarray.shape

(3, 3)

In [27]:
np.array({23,23,12})

array({12, 23}, dtype=object)

In [28]:
zeros = np.zeros((3,4))

In [29]:
zeros

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [30]:
zeros.dtype

dtype('float64')

In [31]:
zeros.size

12

In [32]:
zeros.shape

(3, 4)

In [33]:
rng = np.arange(10)

In [34]:
rng

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [35]:
rng.dtype

dtype('int32')

In [36]:
lspace = np.linspace(1,5,3)

In [37]:
lspace

array([1., 3., 5.])

In [40]:
lspace = np.linspace(0,3.14,50)

In [41]:
lspace

array([0.        , 0.06408163, 0.12816327, 0.1922449 , 0.25632653,
       0.32040816, 0.3844898 , 0.44857143, 0.51265306, 0.57673469,
       0.64081633, 0.70489796, 0.76897959, 0.83306122, 0.89714286,
       0.96122449, 1.02530612, 1.08938776, 1.15346939, 1.21755102,
       1.28163265, 1.34571429, 1.40979592, 1.47387755, 1.53795918,
       1.60204082, 1.66612245, 1.73020408, 1.79428571, 1.85836735,
       1.92244898, 1.98653061, 2.05061224, 2.11469388, 2.17877551,
       2.24285714, 2.30693878, 2.37102041, 2.43510204, 2.49918367,
       2.56326531, 2.62734694, 2.69142857, 2.7555102 , 2.81959184,
       2.88367347, 2.9477551 , 3.01183673, 3.07591837, 3.14      ])

np.linspace(starting element(from),ending element(to),no of elements between them(n))
-->will give equally spaced n elements from start to end 

In [42]:
emp = np.empty((2,4))

In [43]:
emp

array([[0.00000000e+000, 0.00000000e+000, 0.00000000e+000,
        0.00000000e+000],
       [0.00000000e+000, 6.22522714e-321, 2.37663529e-312,
        0.00000000e+000]])

In [44]:
emplike = np.empty_like(lspace)

In [45]:
emplike

array([0.        , 0.06408163, 0.12816327, 0.1922449 , 0.25632653,
       0.32040816, 0.3844898 , 0.44857143, 0.51265306, 0.57673469,
       0.64081633, 0.70489796, 0.76897959, 0.83306122, 0.89714286,
       0.96122449, 1.02530612, 1.08938776, 1.15346939, 1.21755102,
       1.28163265, 1.34571429, 1.40979592, 1.47387755, 1.53795918,
       1.60204082, 1.66612245, 1.73020408, 1.79428571, 1.85836735,
       1.92244898, 1.98653061, 2.05061224, 2.11469388, 2.17877551,
       2.24285714, 2.30693878, 2.37102041, 2.43510204, 2.49918367,
       2.56326531, 2.62734694, 2.69142857, 2.7555102 , 2.81959184,
       2.88367347, 2.9477551 , 3.01183673, 3.07591837, 3.14      ])

give a previous array to empty_like and it will create another array like it which we can use ahead for other purposes

In [47]:
ide = np.identity(3)

In [48]:
ide

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

np.identity(n) gives an identity matrix of nxn

In [49]:
arr = np.arange(100)

In [50]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [51]:
arr.reshape(2,50)

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
        16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,
        32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
        48, 49],
       [50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65,
        66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81,
        82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97,
        98, 99]])

In [52]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [53]:
arr = arr.reshape(10,10)

In [54]:
arr

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
       [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
       [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
       [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
       [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
       [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])

In [55]:
arr.shape

(10, 10)

In [56]:
arr = arr.ravel()

In [57]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [58]:
arr.shape

(100,)

.ravel() coverts any array into 1-D array

In [59]:
x = [[2,4,6],[8,10,12],[30,23,27]]

In [60]:
ar = np.array(x)

In [61]:
ar

array([[ 2,  4,  6],
       [ 8, 10, 12],
       [30, 23, 27]])

In [63]:
ar.sum()

122

In [64]:
ar.sum(axis=0)

array([40, 37, 45])

In [65]:
ar.sum(axis=1)

array([12, 30, 80])

axis =0 --> neeche ki taraf arrow
axis = 1 --> side ki taraf arrow

In [66]:
ar.T

array([[ 2,  8, 30],
       [ 4, 10, 23],
       [ 6, 12, 27]])

In [69]:
ar.flat

In [70]:
for item in ar.flat:
    print(item)

2
4
6
8
10
12
30
23
27


In [71]:
ar.ndim

2

ndim -- no of dimensions 

In [72]:
ar.nbytes

36

In [73]:
ar.argmax()

6

In [74]:
one = np.array([2,6,8,4,23,12])

In [75]:
one.argmax()

4

gives the index of maximum element

In [76]:
one.argsort()

array([0, 3, 1, 2, 5, 4], dtype=int64)

Note -- the commands which dont have paranthesis(for ex- ar.size , ar.shape) are called attributes of array while those which have paranthesis like argmax() argsort() are array functions

In [77]:
ar.argmin()

0

In [78]:
one.argmin()

0

In [79]:
ar.argmax(axis=1)

array([2, 2, 0], dtype=int64)

In [80]:
ar

array([[ 2,  4,  6],
       [ 8, 10, 12],
       [30, 23, 27]])

In [81]:
ar.argmax(axis=0)

array([2, 2, 2], dtype=int64)

In [82]:
ar.argsort()

array([[0, 1, 2],
       [0, 1, 2],
       [1, 2, 0]], dtype=int64)

In [83]:
ar.argsort(axis=0)

array([[0, 0, 0],
       [1, 1, 1],
       [2, 2, 2]], dtype=int64)

In [84]:
ar

array([[ 2,  4,  6],
       [ 8, 10, 12],
       [30, 23, 27]])

In [85]:
ar2 = np.array([[3,1,2],[5,7,1],[12,13,14]])

In [86]:
ar2

array([[ 3,  1,  2],
       [ 5,  7,  1],
       [12, 13, 14]])

In [87]:
ar + ar2

array([[ 5,  5,  8],
       [13, 17, 13],
       [42, 36, 41]])

addition of corresponding indicies of both the matrices

In [88]:
ar * ar2

array([[  6,   4,  12],
       [ 40,  70,  12],
       [360, 299, 378]])

multiplication of corresponding indices of both the matrices

In [89]:
np.sqrt(ar)

array([[1.41421356, 2.        , 2.44948974],
       [2.82842712, 3.16227766, 3.46410162],
       [5.47722558, 4.79583152, 5.19615242]])

In [90]:
ar.sum()

122

In [91]:
ar

array([[ 2,  4,  6],
       [ 8, 10, 12],
       [30, 23, 27]])

In [92]:
ar.max()

30

In [94]:
ar.min()

2

In [95]:
np.where(ar>5)

(array([0, 1, 1, 1, 2, 2, 2], dtype=int64),
 array([2, 0, 1, 2, 0, 1, 2], dtype=int64))

In [96]:
np.count_nonzero(ar)

9

In [97]:
np.nonzero(ar2)

(array([0, 0, 0, 1, 1, 1, 2, 2, 2], dtype=int64),
 array([0, 1, 2, 0, 1, 2, 0, 1, 2], dtype=int64))

In [98]:
ar2

array([[ 3,  1,  2],
       [ 5,  7,  1],
       [12, 13, 14]])

In [99]:
ar2[0,0] = 0

In [100]:
np.nonzero(ar2)

(array([0, 0, 1, 1, 1, 2, 2, 2], dtype=int64),
 array([1, 2, 0, 1, 2, 0, 1, 2], dtype=int64))

Proof that numpy array takes less space them a pyhton list :

In [102]:
import sys

In [103]:
py_ar = [3,5,6,8]

In [104]:
np_ar = np.array(py_ar)

In [105]:
sys.getsizeof(1)*len(py_ar)

112

In [106]:
np_ar.itemsize*np_ar.size

16

numpy array takes only 16 bytes while python list takes 112 bytes

In [108]:
np_ar.tolist()

[3, 5, 6, 8]